File: internal/api/instance.go
package api

import (
	"dockernas/internal/config"
	"dockernas/internal/models"
	"dockernas/internal/service"
	"encoding/json"
	"log"
	"strconv"

	"github.com/gin-gonic/gin"
)

func PostInstance(c *gin.Context) {
	var param models.InstanceParam
	c.BindJSON(&param)
	if config.IsInstanceNameConflict(param.Name) {
		panic(param.Name + " is internal used")
	}
	service.CheckParamIsValid(param)
	service.CreateInstance(param, false)

	c.JSON(200, gin.H{
		"state": "ok",
	})
}

func GetInstance(c *gin.Context) {
	instances := service.GetInstance()
	c.JSON(200, gin.H{
		"list": instances,
	})
}

func GetInstanceByName(c *gin.Context) {
	name := c.Param("name")
	instance := service.GetInstanceByName(name)
	c.JSON(200, instance)
}

func PatchInstance(c *gin.Context) {
	name := c.Param("name")
	instance := models.GetInstanceByName(name)

	patchMap := map[string]string{}
	c.BindJSON(&patchMap)

	op, _ := patchMap["op"]
	if op == "stop" {
		service.StopInstance(*instance)
	}
	if op == "start" {
		service.StartInstance(*instance)
	}
	if op == "edit" {
		data, _ := patchMap["data"]
		var param models.InstanceParam
		err := json.Unmarshal([]byte(data), &param)
		if err != nil {
			log.Println(err)
			panic(err)
		}
		service.CheckParamIsValid(param)
		service.EditInstance(*instance, param)
	}

	c.JSON(200, gin.H{"msg": "ok"})
}

func DeleteInstance(c *gin.Context) {
	name := c.Param("name")
	instance := models.GetInstanceByName(name)
	service.DeleteInstance(*instance)
	c.JSON(200, gin.H{"msg": "ok"})
}

func GetInstanceLog(c *gin.Context) {
	name := c.Param("name")
	instance := models.GetInstanceByName(name)

	log := ""
	if instance.ContainerID != "" && instance.State == models.RUNNING {
		log = service.GetInstanceLog(*instance)
	}

	c.JSON(200, gin.H{"data": log})
}

func GetInstanceEvent(c *gin.Context) {
	name := c.Param("name")
	instance := models.GetInstanceByName(name)
	events := models.GetEvents(instance.Id)
	c.JSON(200, gin.H{"list": events})
}

func GetAllInstanceStats(c *gin.Context) {
	start, err1 := strconv.ParseInt(c.Query("start"), 10, 64)
	end, err2 := strconv.ParseInt(c.Query("end"), 10, 64)

	if err1 != nil {
		panic(err1)
	}
	if err2 != nil {
		panic(err2)
	}

	c.JSON(200, gin.H{"list": models.GetStatDataByTime(start, end)})
}

func GetInstanceStatsByName(c *gin.Context) {
	start, err1 := strconv.ParseInt(c.Query("start"), 10, 64)
	end, err2 := strconv.ParseInt(c.Query("end"), 10, 64)
	name := c.Param("name")

	if err1 != nil {
		panic(err1)
	}
	if err2 != nil {
		panic(err2)
	}

	c.JSON(200, gin.H{"list": models.GetStatDataByName(name, start, end)})
}

File: frontend/src/api/instance.js
import http from '@/utils/request'
import storage from "@/utils/storage"

export const newInstance = (instanceParam) => {
    return http.post("/api/instance", instanceParam)
};

export const getAllInstance = () => {
    return http.get("/api/instance")
}

export const getInstance = (name) => {
    return http.get("/api/instance/" + name)
}

export const stopInstance = (name) => {
    return http.patch("/api/instance/" + name, { "op": "stop" })
}

export const startInstance = (name) => {
    return http.patch("/api/instance/" + name, { "op": "start" })
}

export const editInstance = (name, dataStr) => {
    return http.patch("/api/instance/" + name, { "op": "edit", "data": dataStr })
}

export const deleteInstance = (name) => {
    return http.delete("/api/instance/" + name)
}

export const getInstanceLog = (name) => {
    return http.get("/api/instance/" + name + "/log")
}

export const getInstanceEvent = (name) => {
    return http.get("/api/instance/" + name + "/event")
}

export const getInstanceStats = (start, end) => {
    return http.get("/api/instancestats?start=" + start + "&end=" + end)
}

export const getInstanceStatsByName = (name, start, end) => {
    return http.get("/api/instancestats/" + name + "?start=" + start + "&end=" + end)
}

export const getInstanceHttpPort = () => {
    return http.get("/api/instancehttpport")
}

export const getWebTerminalWebsocket = (containerId, columns) => {
    var url=window.location.host+"/terminal?containerid=" + containerId + "&columns=" + columns + "&token=" + storage.get("token")
    if(window.location.protocol=='http:'){
        url="ws://"+url
    }else{
        url="wss://"+url
    }
    var ws = new WebSocket(url);
    return ws
}
